src/dftbp/dftbplus/mainapi.F90

```diff
@@ -92,6 +96,184 @@ subroutine setGeometry(env, main, coords, latVecs, coordOrigin)
+
+    !> cutoff used for this neighbour list
+    real(dp), intent(in) :: cutOff
+
+    !> coordinates of all images (= atoms in other cells)
+    real(dp), intent(in) :: coord(:,:)
+
+    !> mapping between image index (other cell) and atom index (central cell)
+    integer, intent(in) :: img2CentCell(:)
+
+    real(dp), allocatable :: dist2(:)
+    real(dp), pointer :: rCellVec(:)
+    integer :: nMaxNeighbours, nCellVec, info
+    integer :: iAtom, iCell, iNeigh, iImage, iCellVec, iAtFirst, iAtLast
+    integer, allocatable :: indx(:)
+    logical :: copyData
+
+    if (allocated(main%electronDynamics)) then
+      call error("Not implemented: Cannot set the neighbour list&
+          & when time propagation is enabled")
+    end if
+    if (main%tLocalCurrents) then
+      call error("Not implemented: Cannot set the neighbour list&
+          & when local bond-currents should be computed")
+    end if
+
+    nMaxNeighbours = maxval(nNeighbour)
+
+    main%neighbourList%setExternally = .true.
+
+    #:if WITH_MPI
+      call allocateNeighbourArrays(main%neighbourList, nMaxNeighbours, main%nAtom, .true., env)
+    #:else
+      call allocateNeighbourArrays(main%neighbourList, nMaxNeighbours, main%nAtom, .false.)
+    #:endif
+
+    @:ASSERT(size(nNeighbour) == main%nAtom)
+    main%neighbourList%nNeighbour(:) = nNeighbour(:)
+    main%neighbourList%cutoff = cutOff
+
+    main%nAllAtom = size(img2CentCell) + main%nAtom
+    @:ASSERT(size(img2CentCell) == size(coord, dim=2))
+
+    if (size(main%img2CentCell) /= main%nAllAtom) then
+      call reallocateArrays1(main%img2CentCell, main%iCellVec, main%coord, main%nAllAtom)
+    end if
+
+    !> Prepend data for the atoms in the central cell
+    do iAtom = 1, main%nAtom
+      main%coord(1:3,iAtom) = main%coord0(1:3,iAtom)
+      main%img2CentCell(iAtom) = iAtom
+      main%iCellVec(iAtom) = 1
+    end do
+
+    if (main%nAtom < main%nAllAtom) then
+      main%coord(1:3,main%nAtom+1:) = coord(1:3,:)
+      main%img2CentCell(main%nAtom+1:) = img2CentCell(:)
+      main%iCellVec(main%nAtom+1:) = 0
+
+      !> Now set main%iCellVec: Iterate over all cells, calculate the coordinates the atom would have there,
+      !> and determine the cell in which this is very close to the actual coordinates
+      nCellVec = size(main%rCellVec, dim=2)
+      allocate(dist2(nCellVec))
+
+      rCellVec(1:3*nCellVec) => main%rCellVec
+
+      call distributeRangeInChunks(env, main%nAtom + 1, main%nAllAtom, iAtFirst, iAtLast)
+
+      do iAtom = iAtFirst, iAtLast
+        iImage = main%img2CentCell(iAtom)
+        if (iImage > 0) then
+          do iCell = 1, nCellVec
+            iCellVec = 3 * (iCell - 1)
+            dist2(iCell) = (main%coord0(1,iImage) + rCellVec(iCellVec + 1) - main%coord(1,iAtom))**2 +&
+                & (main%coord0(2,iImage) + rCellVec(iCellVec + 2) - main%coord(2,iAtom))**2 +&
+                & (main%coord0(3,iImage) + rCellVec(iCellVec + 3) - main%coord(3,iAtom))**2
```

Comment:
Yes, you are right, but it turns out that explicitly unrolling the loop by 3 helps the compiler to vectorize this loop. It otherwise -- for some reason -- does not know that there are 3 elements and tries a vectorization over that inner loop.